tests/test_2349_growablebuffer_in_numba.py

```diff
@@ -0,0 +1,508 @@
+    growablebuffer.append(30)
+    assert growablebuffer.snapshot().tolist() == list(range(31))
+    assert len(growablebuffer._panels) == 3
+
+
+def test_python_extend():
+    # small 'initial' and 'resize' for testing
+    growablebuffer = GrowableBuffer(np.int32, initial=10, resize=2.0)
+    assert growablebuffer.snapshot().tolist() == []
+    assert len(growablebuffer._panels) == 1
+
+    # within the first panel
+    growablebuffer.extend(np.array(range(0, 5)))
+    assert growablebuffer.snapshot().tolist() == list(range(5))
+    assert len(growablebuffer._panels) == 1
+
+    # up to (and touching) the end of the first panel (10)
+    growablebuffer.extend(np.array(range(5, 10)))
+    assert growablebuffer.snapshot().tolist() == list(range(10))
+    assert len(growablebuffer._panels) == 1
+
+    # within the second panel (which ends at 30)
+    growablebuffer.extend(np.array(range(10, 20)))
+    assert growablebuffer.snapshot().tolist() == list(range(20))
+    assert len(growablebuffer._panels) == 2
+
+    # touching the end of the second panel
+    growablebuffer.extend(np.array(range(20, 30)))
+    assert growablebuffer.snapshot().tolist() == list(range(30))
+    assert len(growablebuffer._panels) == 2
+
+    # fill one whole panel exactly (start to end)
+    growablebuffer.extend(np.array(range(30, 50)))
+    assert growablebuffer.snapshot().tolist() == list(range(50))
+    assert len(growablebuffer._panels) == 3
+
+    # fill more than one panel, starting at a threshold
+    growablebuffer.extend(np.array(range(50, 80)))
+    assert growablebuffer.snapshot().tolist() == list(range(80))
+    assert len(growablebuffer._panels) == 5
+
+    # fill more than one panel, not starting at a threshold, but ending on one
+    growablebuffer.extend(np.array(range(80, 110)))
+    assert growablebuffer.snapshot().tolist() == list(range(110))
+    assert len(growablebuffer._panels) == 6
+
+    # fill lots of panels, starting at a threshold
+    growablebuffer.extend(np.array(range(110, 160)))
+    assert growablebuffer.snapshot().tolist() == list(range(160))
+    assert len(growablebuffer._panels) == 9
+
+    # fill lots of panels, not starting at a threshold or ending on one
+    growablebuffer.extend(np.array(range(160, 200)))
+    assert growablebuffer.snapshot().tolist() == list(range(200))
+    assert len(growablebuffer._panels) == 11
+
+    # fill lots of panels, not starting at a threshold, but ending on one
+    growablebuffer.extend(np.array(range(200, 250)))
+    assert growablebuffer.snapshot().tolist() == list(range(250))
+    assert len(growablebuffer._panels) == 13
+
+    # fill a whole lot of panels, just for fun
+    growablebuffer.extend(np.array(range(250, 1000)))
+    assert growablebuffer.snapshot().tolist() == list(range(1000))
+    assert len(growablebuffer._panels) == 51
+
+
+def test_unbox():
+    @numba.njit
+    def f1(x):
+        x
+        return 3.14
+
+    growablebuffer = GrowableBuffer(np.int32)
+    f1(growablebuffer)
```

Comment:
```suggestion
    assert f1(growablebuffer) == 3.14
```

Reply:
Thanks for the suggestion, but this function returns 3.14 just so that it can return something. Checking its value doesn't test anything related to the GrowableBuffer implementation itself. Just calling `f1` is enough to verify that it doesn't segfault.

What I really wanted to do here is call `f1` repeatedly in a loop and ensure that the `PyObject`s in the GrowableBuffer Python implementation (the `numba.typed.List` and the NumPy arrays) don't leak. I did that test offline, but when I tried to incorporate it into the test suite, I was getting extra references. They weren't growing with the number of iterations (an indication of a leak), but they differed from the test values and couldn't be asserted. I know that pytest rewrites the test code, to instrument it with better diagnostics, and perhaps it was rewriting it in a way that acquired additional references.

Since you'll be writing more Numba code, we should keep this in mind, that we need to find a better way to check Python reference counts. Moreover, just checking them on the Python side is not always enough, since some Python objects are created in Numba functions, such as boxing and unboxing functions. Since we don't have any reference to those objects in Python, there's no way to call `sys.getrefcount` on them. CPython has an—undocumented!—Py_RefCnt function, but it's not exposed through Numba's `c.pyapi`.

It's also the case that a Python reference count is the first object in a `PyObject*`, so it can be accessed by unsafe casting. I did that horrible thing here: https://github.com/jpivarski-talks/2019-04-08-picscie-numpy/blob/master/09-jailbreaking-python.ipynb

```python
# The snake eats its own tail: reading PyObject refcount without changing it.
class PyObject(ctypes.Structure): pass
PyObject._fields_ = [
    ("ob_refcnt", ctypes.c_size_t),
    ("ob_type", ctypes.POINTER(PyObject)),
    # there are more struct fields; different Python types have a different number of them
]

hello = b"Hello, world!"
ptr = PyObject.from_address(id(hello))
print("ptr =", ptr)

# a few reference counts
print("\nBefore:")
print("ptr.ob_refcnt =", ptr.ob_refcnt, "sys.getrefcount(hello) =", sys.getrefcount(hello))

# a lot more reference counts
biglist = [hello] * 1000
print("\nAfter:")
print("ptr.ob_refcnt =", ptr.ob_refcnt, "sys.getrefcount(hello) =", sys.getrefcount(hello))
```

Oh, but about this suggested change: I'm going to not take it, so that I can merge the PR now.